encapsulate implicit state

bryevdv opened the pull request:
issues: closes #1916

bryevdv commented:
@bokeh/core I'd like to get some preliminary feedback. Since this is such an important interface, I do not intend to merge it until there is 100% test coverage, as well as docs. But I wanted to see if my usage of mock and patch seemed reasonable.

bryevdv commented:
I'd also like to make this PR with docs and tests a "prototypical" model for PRs as we go forward. We need to dramatically increase test coverage as well as improve our docs, before 0.9 release.

bryevdv commented:
@mattpap I'd also like your thoughts on `_push_or_save` and `_dedupicate_plots`. I think they should be moved, and perhaps made public as well. But alternatives or other ideas are welcome.

bryevdv commented:
So the tests for `show` and `save` etc have tons of mocks. But these functions really just delegate to other things (that happen to have lots of side effects) so I think this is ok, basically just trying to crystallize the expected operations. Again, thoughts welcome. If there is a better way to test these I would like to know.

bryevdv commented:
I updated all the charts examples to all use `output_file` and `output_notebook` and `show(plot)` instead of `filename="foo.html"` and `notebook=True` and `.show()`. It's a pretty huge documentation burden to have to maintain two explanations (and also leads to questions and confusion) so I think we should phase those out completely, eventually. I also updated all the examples for style consistency and correct labeling (several plot titles were wrong).

bryevdv commented:
@damianavila @mattpap @fpliger @almarklein I am going to try and finish adding the tests today. I would appreciate any feedback about this PR. The intent is that this is the "easy" interface for people to use to push, save, or show plots an layouts, regardless of API used for plotting.

fpliger commented:
Wow, completely forgot about this one during the big rush for release. Sorry. This is a great addition, will take a look today.

damianavila commented:
It is in my todo list for today make a review :wink:

almarklein commented on examples/charts/steps.py:
```diff
@@ -1,9 +1,6 @@
-import pandas as pd
-from bokeh.charts import Chart, Step, Line
-from bokeh.plotting import show,output_file
+from bokeh.charts import Step, show, output_file
```
Does this mean there is no more `plotting.show()`?

almarklein replied:
Forget it. I see `show` gets imported from state.

bryevdv replied:
Ideally users would import `show`, etc. from state, but we now have its of code that imports it from `plotting`. As a compromise I'd rather have both `plotting` and `charts` import `show` etc, and then users can use `show` from either. The main point is to avoid "mixing" the APIs unnecessarily. If you just want a simple chart its confusing to users to have to import things from the `plotting.py` API as well. Eventually maybe we can deprecate the usage from `plotting.py` and `charts.py` and move everyone to importing from `state.py`

almarklein replied:
FWIW. I quite like to be able to do `from bokeh.charts import Step, show`, instead of needing an extra line just to import `show`.

almarklein replied:
I guess this is "practicality beats purity" :P

damianavila replied:
> FWIW. I quite like to be able to do from bokeh.charts import Step, show, instead of needing an extra line just to import show.

I see some value in your opinion, but I also think that the unified approach in `state` have conceptually gains over a "distributed" model where you can use `show` from whatever API...

bryevdv replied:
I'm not sure I have an opinion beyond strongly wanting to prevent
```
from bokeh.plotting import show
from bokeh.charts import Line
```

damianavila replied:
I guess we can wait for our users to choose... if we see they use the `show` in each API, then we can leave that for them... but we have to choose a way for our examples, and I guess using the `state` is the correct one...

fpliger replied:
I think I have a [no strong] preference to be able to do: 

```Python
from bokeh.plotting import show, figure
# or
from bokeh.charts import Line, show
# without have to always 
from bokeh.state import show
```
even if all `show` are the same method. Repeating is not a good pattern but in this case I see value in having it. (maybe I'm too lazy? :stuck_out_tongue_closed_eyes: )

almarklein commented on bokeh/state.py:
```diff
@@ -0,0 +1,490 @@
+        title = state.file['title']
+
+    if not filename:
+        warnings.warn("save() called but no filename was supplied and output_file(...) was never called, nothing saved")
```
Just wondering, since I saw this elsewhere. When should we use `warning.warn()` and when `logging.warn`? What I saw is that `warning.warn` does not pass through `bokeh.logging`.

bryevdv replied:
I actually changed this to raise an exception.

bryevdv commented:
BTW is `state` the right name for this module? More precisely it might be described as "output state" but I like the terseness of `state`.

bryevdv commented:
OK, I think I am about done with this. Still would like some input on what to do with `_push_or_save` and `_deduplicate_plots`

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+
+# Third-party imports
+
+# Our own imports
```
Bokeh imports :wink:

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+        document :
+        file :
+        notebook :
+        session :
```
I guess this is will be completed later...

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+                command). If False, then the file is only saved upon calling
+                show().
+
+            mode (str, optional) : how to inlcude BokehJS (default: "inline")
```
`include`

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+
+        if clear:
+            self._document.clear()
+
```
OK, you have the new method: `output_file`, `output_notebook` and `output_server`... do you want to document it in the same way you do it outside the class? Maybe a simple docstring is enough and you get the full documentation in the functions below... or even better... maybe you can, in some way, documents these methods and make the functions replicate the documentation from the methods... if not there is a duplication of information an both docs can be unsynced in the future...

bryevdv replied:
Maybe these methods should be private, I'm not sure. I guess the question is: do we ever expect it to be useful for anyone to create their own `State` objects.

fpliger replied:
I don't see clear use cases for this now but I'd bet my 2c (ok, let's bet a :beer: ) that it'll pop up soon.

damianavila replied:
ok... if we bet with @fpliger... what we don then?

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+            command). If False, then the file is only saved upon calling
+            show().
+
+        mode (str, optional) : how to inlcude BokehJS (default: "inline")
```
`include`

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+    cell. Otherwise, a browser window or tab is autoraised to display the
+    plot object.
+
+    If a both a server session and notebook output have been configured on
```
delete `a`

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+
+def _show_server_with_state(obj, state, new, controller):
+    push(state=state)
+    controller.open(state.session.object_link(state.document.context), new=_new_param[new])
```
I like overall this stacked way for `show` , :+1:

fpliger replied:
very nice indeed!

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,516 @@
+        title = state.file['title']
+
+    if filename is None:
+        raise RuntimeError("save() called but no filename was supplied and output_file(...) was never called, nothing saved")
```
wondering if we don't have to provide a default `filename` here...

bryevdv replied:
That seems like a bad idea, the last thing I want is bug reports of us accidentally overwriting someone's important file with a default filename that gets used implicitly.

damianavila replied:
ok, fair enough... I missed that issue...

fpliger commented on bokeh/state.py:
```diff
@@ -0,0 +1,568 @@
+def _push_or_save(obj):
+    if _state.session and _state.document.autostore:
+        push()
+    if _state.file and _state.file['autosave']:
```
I wonder... should the higher level `open`/`push`/etc... methods live here or somewhere else? I mean... I can see more then this verically shared by all the `bokeh` APIs (see palette, tools and other things..). Should those higher level things live here or in a `shared` package that exposed all those?

fpliger commented on bokeh/tests/test_plotting.py:
```diff
@@ -124,10 +114,8 @@ def test_grid(self):
-        self.assertEqual(plt._default_file['resources'].minified, True)
-        plt.reset_output()
+    def test_legend(self):
+        pass
```
why this?

fpliger commented on bokeh/tests/test_plotting.py:
```diff
@@ -138,5 +126,10 @@ def test_tools(self):
+class TestMarkers(unittest.TestCase):
+
+    def test(self):
+        pass
```
same comment as above..

damianavila replied:
I guess it is a place holder...

bryevdv replied:
yup

fpliger commented on examples/charts/area.py:
```diff
@@ -1,6 +1,7 @@
-from bokeh.charts import Area
-from bokeh.plotting import show, output_file
+
+from bokeh.charts import Area, show, output_file
```
nice to read this... :+1:

damianavila replied:
so, you want to encourage the use of the transitive import? If I have to choose, I will promote importing from `state`...

fpliger replied:
well.. it we are planning on having `show` and friends only on `state` in a future release then I agree with you. We should promote it from now..

bryevdv replied:
I'll confess to being torn on this one. It's really, really nice from  user's perspective to just have all the needed tools in one place. I am not really sure there is any negative consequence for us on the dev side. I would say that maybe we do move the functions to their own "io" api module, and make a point to always import the entire "io" api module into the other interfaces.

damianavila replied:
I like the centralize thing... but as I said in other PR, I also see value in this... mainly from the pragmatic perspective...

>I would say that maybe we do move the functions to their own "io" api module, and make a point to always import the entire "io" api module into the other interfaces.

Yep...

damianavila commented on bokeh/state.py:
```diff
@@ -0,0 +1,568 @@
+    '''
+    _state.reset()
+
+def _deduplicate_plots(plot, subplots):
```
about this two... why not move vbox and hbox (from `selection_histogram.py`) here??

bryevdv replied:
Quite possibly. I really want to get @mattpap opinion on these. It might make sense to move all the functions, and the global `_state` plus some things like `grid`, `vbox` and `hbox` to a shared module as you describe below.

damianavila commented:
well, I left some comments... in general, this look good to me...

fpliger commented:
> I added all the bokeh.state functions drectly to bokeh.charts and I updated all the charts examples to all use output_file and output_notebook and show(plot) (from bokeh.charts) instead of filename="foo.html" and notebook=True and .show(). It's a pretty huge documentation burden to have to maintain two explanations (and also leads to questions and confusion) so I think we should phase those out completely

although I like `.show` I agree it a documentation burden.. +1 for consistency on all levels

fpliger commented:
> Still would like some input on what to do with _push_or_save and _deduplicate_plots

Why not make them public?

bryevdv commented:
> Why not make them public?

They are pretty esoteric. I think the idea of making `vbox`, `hbox` and `grid` (the things that use them) public is a better idea.

bryevdv commented:
Proposal then:
```
bokeh.state:
 * State 

bokeh.io:
 * output_file
 * output_notebook
 * output_server
 * push
 * save 
 * show
 * vbox
 * hbox
 * grid (or gridplot... I think we should rename gridplot to this, with a deprecation)
```

damianavila commented:
> Proposal then:

Sounds good to me...

fpliger commented:
Sounds good to me. Really like the `bokeh.io` name.

bryevdv commented:
@damianavila @fpliger @mattpap OK I made the split. I like `bokeh.io` as well. Also I moved the "vbox" and "hbox" functions into the new module and renamed then `vplot` and `hplot` which matches `gridplot` nicely (also moved to `bokeh.io)

fpliger commented:
:+1: LGTM

bryevdv commented:
further proposal: move `bokeh.state` to `bokeh.utils.state` (when we split up `utils.py` into a package)

bryevdv commented:
OK, there was one unit test failure form the re-org, I fixed that. Tests should pass now. @fpliger @damianavila @mattpap I intend to merge this tomorrow morning CST.

damianavila commented:
LGTM

bryevdv commented:
@bokeh/core OK, merging this, we can continue to tweak/improve. I am interested in thoughts about the `hplot` and `vplot`, etc. @mattpap if we we're to move to this as the standard way to do things in non-models code, we could probably remove the initializer conveniences in `VBox` and `HBox` eventually.
